### PR TITLE
support large scale for ads mode

### PR DESCRIPTION
**Pull request by nlgwcy**
**What type of PR is this?**

/kind enhancement

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

File: bpf/deserialization_to_bpf_map/deserialization_to_bpf_map.c [hunk 1]
```diff
@@ -541,7 +575,7 @@ static int map_info_check(struct bpf_map_info *outter_info, struct bpf_map_info
 
     if (outter_info->max_entries < 2 || outter_info->max_entries > MAX_OUTTER_MAP_ENTRIES) {
         LOG_ERR("outter map max_entries must be in[2,%d]\n", MAX_OUTTER_MAP_ENTRIES);
-        return -EINVAL;
+        // return -EINVAL;
```
Comment (hzxuzhonghu) [inline bpf/deserialization_to_bpf_map/deserialization_to_bpf_map.c:578]: Should we update the if check now? IIUC The max_entries can exceed MAX_OUTTER_MAP_ENTRIES after scale out

Comment (hzxuzhonghu) [reply bpf/deserialization_to_bpf_map/deserialization_to_bpf_map.c:578]: So should this be equal to MAX_OUTTER_MAP_ENTRIES

File: bpf/deserialization_to_bpf_map/deserialization_to_bpf_map.c [hunk 2]
```diff
@@ -1244,29 +1278,64 @@ int get_outer_inner_map_infos(
...
-        if (!g_inner_map_mng.inner_maps[i].map_fd) {
+        idx = g_inner_map_mng.elastic_slots[i];
+        if (!g_inner_map_mng.inner_maps[idx].map_fd) {
+            continue;
+        }
+
+        ret = bpf_map_update_elem(ctx->outter_fd, &idx, &g_inner_map_mng.inner_maps[idx].map_fd, BPF_ANY);
+        if (ret)
+            break;
+        g_inner_map_mng.inner_maps[idx].alloced = 1;
```
Comment (hzxuzhonghu) [inline bpf/deserialization_to_bpf_map/deserialization_to_bpf_map.c:1297]: Is thre a lock to prevent conflict. Otherwise it is not safe

**Review by hzxuzhonghu — COMMENTED** [review]
should we consider inner map r/w conflict

File: bpf/deserialization_to_bpf_map/deserialization_to_bpf_map.c (hunk 1 above)
Comment (nlgwcy) [reply bpf/deserialization_to_bpf_map/deserialization_to_bpf_map.c:578]:
```c
struct {
    __uint(type, BPF_MAP_TYPE_ARRAY_OF_MAPS);
    __uint(key_size, sizeof(__u32));
    __uint(value_size, sizeof(__u32));
    __uint(max_entries, MAP_SIZE_OF_OUTTER_MAP);
    __uint(map_flags, 0);
} outer_map SEC(".maps");
```

The maximum number of records has been specified when outter_map is defined. Here, only verification is performed.

File: bpf/deserialization_to_bpf_map/deserialization_to_bpf_map.c [hunk 3]
```diff
@@ -51,23 +52,48 @@ struct op_context {
...
 };
 
 struct task_contex {
     int outter_fd;
     int task_id;
+    bool scaleup;
+};
+
+#define MAP_IN_MAP_MNG_PERSIST_FILE_PATH "/mnt/mim_mng_persist"
+#define MAGIC_NUMBER                     0xb809b8c3
```
[the remark this answers is not in the document]
Comment (nlgwcy) [reply bpf/deserialization_to_bpf_map/deserialization_to_bpf_map.c:84]: This is a pre-embedded feature. After the configuration restoration feature is supported, the management information of the inner_map is made persistent to a file. Upon the next startup, the validity of the management information is simply verified based on magic.

**Review by hzxuzhonghu — DISMISSED** [review]

**Comment by hzxuzhonghu** [comment]
TestRestartReload Still failed

File: bpf/deserialization_to_bpf_map/deserialization_to_bpf_map.c [hunk 4]
```diff
@@ -1603,6 +1603,8 @@ void deserial_uninit(bool persist)
 {
     if (persist)
         inner_map_mng_persist();
+    else
+        remove(MAP_IN_MAP_MNG_PERSIST_FILE_PATH);
```
Comment (hzxuzhonghu) [inline bpf/deserialization_to_bpf_map/deserialization_to_bpf_map.c:1607]: hha

File: pkg/utils/test/bpf_map.go [hunk 5]
```diff
@@ -31,6 +31,7 @@ import (
 type CleanupFn func()
 
 func InitBpfMap(t *testing.T, config options.BpfConfig) (CleanupFn, *bpf.BpfLoader) {
+	bpf.CleanupBpfMap()
```
Comment (hzxuzhonghu) [inline pkg/utils/test/bpf_map.go:34]: why do we need this? Any test forget to cleanup before exit?

Comment (nlgwcy) [reply pkg/utils/test/bpf_map.go:34]: sorry... I forget to remove, just for test.

**Pull request merged**